File: summer_2024/OOP - Diffie-Hellman/Java/DHSetup.java
import java.util.ArrayList;
import java.security.SecureRandom;

public class DHSetup<T extends Field> {
    
    private T generator;
    private Class<T> fieldType;

    private ArrayList<Long> sieveOfEratosthenes(int n){

        ArrayList<Long> primes = new ArrayList<>();
        boolean prime[] = new boolean[n+2];
        prime[0] = false;
        prime[1] = false;

        for (int p = 2; p * p <= n; p++) {
            if (prime[p]) {
                for (int i = p * p; i <= n; i += p)
                    prime[i] = false;
            }
        }
    
        for (int p = 2; p <= n; p++) {
            if (prime[p])
                primes.add((long) p);
        }
    
        return primes;

    }

    public DHSetup(Class<T> fieldType) {
        this.fieldType = fieldType;
        SecureRandom rd = new SecureRandom();
        boolean found = false;
        while (!found) {
            int chosen = rd.nextInt((int) (Field.getCharacteristic() - 2)) + 2;
            generator = createInstance(chosen);
            ArrayList<Long> primesSmallerThanP = sieveOfEratosthenes(chosen);
            boolean isGenerator = true;
    
            for (long i : primesSmallerThanP) {
                if ((Field.getCharacteristic() - 1) % i == 0) {
                    if (power(generator, (Field.getCharacteristic() - 1) / i).getValue() == 1) {
                        isGenerator = false;
                        break;
                    }
                }
            }
    
            if (isGenerator) {
                generator = createInstance(chosen);
                found = true;
            }
        }
    }

    private T createInstance(int value) {
        try {
            return fieldType.getDeclaredConstructor(long.class).newInstance((long) value);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    public T getGenerator(){
        return generator;
    }

    public T power(T a, long b){
        T res = a;
        while(b>0){
            if(b%2 == 1){
                res.assign(res.mul(a));
            }
            a.assign(a.mul(a));
            b /= 2;
        }
        return res;
    }
}

File: summer_2024/OOP - Diffie-Hellman/Python/DHSetup.py
from typing import List, TypeVar, Generic, Type
from Field import P
from Field import Field
import random

T = TypeVar('T')
class DHSetup(Generic[T]):

    def sieveOfEratosthenes(self, n):

        primes = []

        prime = [True for i in range(n+1)]
        p = 2
        while (p*p <= n):
            if(prime[p == True]):
                for i in range(p * p, n+1, p):
                    prime[i] = False
            p += 1

        for p in range(2, n+1):
            if prime[p]:
                primes.append(p)

        return primes
    
    def __init__(self, member_class: Type[T]):
        found = False
        self.generator = None

        while(not found):
            
            chosen = random.randint(2,P-1)
            self.generator = member_class(chosen)
            primesSmallerThanP = self.sieveOfEratosthenes(chosen)
            isGenerator = True


            for i in range(0,len(primesSmallerThanP)):
                if (P-1) % primesSmallerThanP[i] == 0:
                    if self.power(self.generator, ((P-1)/primesSmallerThanP[i])) == Field(1):
                        isGenerator = False
                        break
            
            if isGenerator:
                self.generator = member_class(chosen) #???
                found = True

        
    def getGenerator(self) -> T:
        return self.generator
    
    def power(self, a: T, b) -> T:
        res = a
        while b>0:
            if b%2 == 1:
                res = res*a
            a = a*a
            b /= 2

        return res
    